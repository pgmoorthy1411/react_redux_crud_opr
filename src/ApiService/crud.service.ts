class ApiService {

    getData() {
        return fetch('https://jsonplaceholder.typicode.com/users')
            .then(response => response.json())
    }

    postData = async (postData: any) => {
        try {

            const response = await fetch('https://jsonplaceholder.typicode.com/users', {
                method: 'post',
                headers: {
                    'Accept': 'application/json, text/plain, */*',
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(postData)
            });
            const data = await response.json();
            return data;
        } catch (error) {
            return error;

        }
    }

    editData = async (postData: any) => {
        return fetch(`https://jsonplaceholder.typicode.com/users/${postData}`)
            .then(response => response.json())
    }

    updateData = async (postData: any) => {
        try {

            const response = await fetch(`https://jsonplaceholder.typicode.com/users`, {
                method: 'put',
                headers: {
                    'Accept': 'application/json, text/plain, */*',
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(postData)
            });
            const data = await response.json();
            return data;
        } catch (error) {
            return error;

        }
    }
    deleteData = async (postData: any) => {
        try {

            const response = await fetch(`https://jsonplaceholder.typicode.com/users`, {
                method: 'delete',
                headers: {
                    'Accept': 'application/json, text/plain, */*',
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify(postData)
            });
            const data = await response.json();
            return data;
        } catch (error) {
            return error;

        }
    }

}

export default new ApiService();