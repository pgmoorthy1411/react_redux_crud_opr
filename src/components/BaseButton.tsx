import React from 'react';
import {Button} from '@mui/material';

const BaseButton = ({ variant, color,  text} : any) => {

  return <Button variant={variant} color={color} title='Submit Buttons'  type="submit" >{text}</Button>  ;
};

export default BaseButton;
