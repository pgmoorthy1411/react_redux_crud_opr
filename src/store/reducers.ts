import { combineReducers } from 'redux';
import crudReducer from './crud/reducer';

const rootReducers = combineReducers({
    crudReducer
});

export default rootReducers;