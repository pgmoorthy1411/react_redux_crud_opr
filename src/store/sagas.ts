import crudSaga from "./crud/saga";
import { all, fork } from 'redux-saga/effects';

export default function* rootSaga(){
    yield all([
        fork(crudSaga)
    ])
}