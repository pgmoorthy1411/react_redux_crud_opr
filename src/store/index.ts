import { createStore, applyMiddleware, compose } from 'redux';
import createSagaMiddleware from "redux-saga";
import reducers from './reducers';
import rootSaga from "./sagas";

const sagaMiddleware = createSagaMiddleware();
const middlewares = [sagaMiddleware];

const composeEnhancers = compose();

export function configureStore(initialState: any) {

    const store = createStore(
        reducers,
        initialState,
        composeEnhancers(
            applyMiddleware(...middlewares)
        ),
    );
    sagaMiddleware.run(rootSaga);
    return store;
}

