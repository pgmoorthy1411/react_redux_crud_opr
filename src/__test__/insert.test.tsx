import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Insert from '../pages/insert';
import { BrowserRouter, Route, Routes } from 'react-router-dom';
import { configureStore } from '../store';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import { Button } from '@mui/material';
import BaseButton from '../components/BaseButton';
import ReactDOM from 'react-dom';
import ApiService from '../ApiService/crud.service';
jest.mock('../ApiService/crud.service');


test('Insert Page Render Test', () => {

  const insertModule = render(
    <Provider store={configureStore({})}>
      <BrowserRouter >
        <div className="App">
          <Routes>
            <Route path="/" element={<Insert />} />
          </Routes>
        </div>
      </BrowserRouter>
    </Provider>
  );
  // console.log(insertModule);
  expect(insertModule.getByText('Add User')).toBeInTheDocument();
  // expect(insertModule.queryByTitle('Submit Buttons')).toBeInTheDocument();
  // const submitBtn = insertModule.queryByTitle('Submit Buttons');
  // expect(submitBtn).toBeTruthy();

});

describe('Testing  button in Insert Page', () => {

  it('onclick', () => {

    const insertModule = render(
      <Provider store={configureStore({})}>
        <BrowserRouter >
          <div className="App">
            <Routes>
              <Route path="/" element={<Insert />} />
            </Routes>
          </div>
        </BrowserRouter>
      </Provider>
    );
    expect(insertModule.getByText('Submit')).toBeInTheDocument();
    const submitBtn = insertModule.getByText('Submit');
    expect(submitBtn).toBeTruthy();
    expect(submitBtn).toMatchSnapshot();

    // fireEvent.click(submitBtn)
    // console.log(submitBtn.innerHTML);

    // expect(submitBtn.innerHTML).toBe("Submit");

  })

});



describe('Form Test in Insert Page', () => {


  let container: HTMLDivElement;
  const insertApiServiceSpy = jest.spyOn(ApiService, 'postData');
  const getApiServiceSpy = jest.spyOn(ApiService, 'getData');

  beforeEach(() => {
    container = document.createElement('div');
    document.body.append(container);
    ReactDOM.render(
      <Provider store={configureStore({})}>
        <BrowserRouter >
          <div className="App">
            <Routes>
              <Route path="/" element={<Insert />} />
            </Routes>
          </div>
        </BrowserRouter>
      </Provider>,
      container
    )
  });

  afterEach(() => {
    document.body.removeChild(container);
    container.remove();
  })

  it('Form renders', () => {
    const inputs = container.querySelectorAll('input')
    console.log(inputs);
    console.log(inputs.length);
    expect(inputs.length).toBe(3);
    expect(inputs[0].name).toBe('firstName');
    expect(inputs[1].name).toBe('email');
    expect(inputs[2].name).toBe('phone');
  })


  describe("mockImplementation", () => {
    test("function", () => {
      const mockFn1 = jest.fn().mockImplementation(() => 42);
      const mockFn2 = jest.fn(() => 42);

      expect(mockFn1()).toBe(42);
      expect(mockFn2()).toBe(42);
    });
  });

  // 
  it('Values check', () => {


    const inputs = container.querySelectorAll('input')
    const name = inputs[0];
    const email = inputs[1];
    const phone = inputs[2];
    const submitBtn = container.querySelectorAll('button[type=button]')
    console.log(submitBtn.length);
    // expect(submitBtn[0]).toBeTruthy();
    let getName = fireEvent.change(name, { target: { value: 'firstName' } })
    let getEmail = fireEvent.change(email, { target: { value: 'email' } })
    let getPhone = fireEvent.change(phone, { target: { value: '9876543210' } })

    const fakeUserResponse = {
      'data': {
        email: "jcmoorthy5050@gmail.com",
        firstName: "pgmoorthy",
        phone: "9876543210"
      }
    };
    // expect(container.getByText('Submit')).toBeInTheDocument();

    fireEvent.click(submitBtn[0])


    var apiFunc = jest.spyOn(ApiService, 'postData').mockImplementationOnce(() => {
      return Promise.resolve({
        json: () => Promise.resolve(fakeUserResponse)
      })
    })
    // console.log('====================================');
    // console.log(apiFunc);
    // console.log('====================================');
    // const submitBtnInsert = container.getByText('Submit');


    // console.log('firstName', getName);
    // console.log('email', getEmail);
    // console.log('phone', getPhone);

    //  let data = {
    //   names: getName,
    //   emails: getEmail,
    //   phones: getPhone
    // }

    // let data = {
    //   email: "jcmoorthy5050@gmail.com",
    //   firstName: "pgmoorthy",
    //   phone: "9876543210"
    // }

    // console.log(data);

    // expect(insertApiServiceSpy).toBeCalledWith(data ,expect.any(Object), expect.any(Function));
    // expect(insertApiServiceSpy).toBeCalledWith(getName, getEmail, getPhone);
    // expect(insertApiServiceSpy).toBeCalledWith(
    //   {
    //     firstName: getName, email: getEmail,phone: getPhone
    //   },expect.any(Object), expect.any(Function)
    //  );

    // expect(insertApiServiceSpy).toBe(data);

  })



});

test('button test', () => {

  const insertModule = render(
    <Provider store={configureStore({})}>
      <BrowserRouter >
        <div className="App">
          <Routes>
            <Route path="/" element={<Insert />} />
          </Routes>
        </div>
      </BrowserRouter>
    </Provider>
  );
  expect(insertModule.getByText('Submit')).toBeInTheDocument();
  const submitBtn = insertModule.getByText('Submit');
  // console.log(submitBtn)
  expect(submitBtn).toBeTruthy();

  const baseBtn = render(<BaseButton text="Submit" variant="contained" />)
  // console.log(baseBtn)
  expect(baseBtn).toBeTruthy();
  // expect(submitBtn).toBE(baseBtn);
  expect(submitBtn).toMatchSnapshot();

  // const wrapper = renderer.create( <BaseButton text="Submit" variant="contained" />);
  // // console.log(wrapper.toJSON())
  // expect(wrapper.toJSON()).toBe(submitBtn);
})



// describe("post requests",()=>{
  //   // const validRequest={method:'post',body:{user:"Jill",password:'hill',post:"He broke his crown!"}}
  //   const validRequest={
  //     email: "jcmoorthy5050@gmail.com",
  //     firstName: "pgmoorthy",
  //     phone: "9876543210"
  //   }
  //   const invalidRequest={method:'post',body:{user:"Jill",password:'beanstock',post:"Jack is cool..."}}

  //   it("returns a 200 status and adds the post to the database",()=>{
  //    const newDatabase={
  //    users:[
  //     {
  //       name:"Jack",
  //       passwordHash:"dasdKDKDJSLASDLASDJSAasdsdc123",
  //       posts:["I just bought some magic beans!"]
  //     },
  //     {
  //       name:"Jill",
  //       passwordHash:"dasdKDKDJSLASDLASDJSAasdsdc123",
  //       posts:["Jack fell down!","He broke his crown!"]
  //     },
  //    ]
  //    }
  //    const mockApiCall=ApiService.postData(validRequest)
  //    console.log('====================================');
  //    console.log(mockApiCall);
  //    console.log('====================================');

  //   //  return mockApiCall.then(response=>{
  //   //  expect(response.status).toBe(200)
  //   // //  expect(ApiService.ApiService).toEqual(newDatabase)
  //   //  })
  //   })
  // })
