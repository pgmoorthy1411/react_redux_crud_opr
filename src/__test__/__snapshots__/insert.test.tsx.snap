// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing  button in Insert Page onclick 1`] = `
<button
  class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  title="Submit Buttons"
  type="submit"
>
  Submit
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;

exports[`button test 1`] = `
<button
  class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
  tabindex="0"
  title="Submit Buttons"
  type="submit"
>
  Submit
  <span
    class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
  />
</button>
`;
